proxy/src/transparency/tcp.rs

```diff
@@ -71,14 +73,177 @@ impl Proxy {
         let fut = connect.connect()
             .map_err(|e| debug!("tcp connect error: {:?}", e))
             .and_then(move |tcp_out| {
-                let (in_r, in_w) = tcp_in.split();
-                let (out_r, out_w) = tcp_out.split();
-
-                copy(in_r, out_w)
-                    .join(copy(out_r, in_w))
-                    .map(|_| ())
+                Duplex::new(tcp_in, tcp_out)
                     .map_err(|e| debug!("tcp error: {}", e))
             });
         Box::new(fut)
     }
 }
+
+/// A future piping data bi-directionally to In and Out.
+struct Duplex<In, Out> {
```

Comment:
It seems to me that this Duplex struct should be in its own module. If it can't be then let's document why not.

Reply:
It *could* be, but it's basically the entirety of this `tcp` module. Splitting it into another isn't really useful, unless we want to make use of `Duplex` in some completely unrelated code. But I feel that if we do want to do that, a refactor can happen at that time.